### PR TITLE
Add parameters for high pT particle protection in PUPPI [106X backport]

### PR DESCRIPTION
**Pull request by ahinzmann**
Backport of #28558

#### PR description:

Add new parameter to PUPPI to deal with high pT particles. Recent studies have shown that the current handling of high momentum charged particles in PUPPI is suboptimal compared to CHS and a different configuration of PUPPI for Run3 is proposed.
https://indico.cern.ch/event/862332/contributions/3632864/
This PR implements the necessary switches without changing the current behavior of the algorithm, i.e. no changes to RECO+MiniAOD are expected from this PR.

At the same time the behavior of the parameter puppiNoLeptons, that is not used anywhere in CMSSW, is adapted such that implementation actually matches the name.

#### PR validation:

runtests

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: CommonTools/PileupAlgos/interface/PuppiContainer.h [hunk 1]
```diff
@@ -52,6 +52,7 @@ class PuppiContainer{
     double fNeutralSlope;
     double fPuppiWeightCut;
     double fPtMax;
+  double fPtMaxNeutralsStartSlope;
```
Comment (perrotta) [inline CommonTools/PileupAlgos/interface/PuppiContainer.h]: Please indent correctly

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 2]
```diff
@@ -145,13 +148,17 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
-        if (tmpFromPV == 3){ pReco.id = 1; }
-        if (tmpFromPV == 1 || tmpFromPV == 2){ 
+        else if (tmpFromPV == 3){ pReco.id = 1; }
+        else if (tmpFromPV == 1 || tmpFromPV == 2){ 
           pReco.id = 0;
-          if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) < fDZCut)) pReco.id = 1;
-          if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) > fDZCut)) pReco.id = 2;
-          if (fPuppiForLeptons && tmpFromPV == 1) pReco.id = 2;
-          if (fPuppiForLeptons && tmpFromPV == 2) pReco.id = 1;
+            if ((fPtMaxCharged > 0) and (pReco.pt > fPtMaxCharged))
```
Comment (perrotta) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: Please indent consistently

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 3]
```diff
@@ -165,13 +172,17 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
-        if (lPack->fromPV() == (pat::PackedCandidate::PVUsedInFit)){ pReco.id = 1; }
-        if (lPack->fromPV() == (pat::PackedCandidate::PVTight) || lPack->fromPV() == (pat::PackedCandidate::PVLoose)){ 
+        else if (lPack->fromPV() == (pat::PackedCandidate::PVUsedInFit)){ pReco.id = 1; }
+        else if (lPack->fromPV() == (pat::PackedCandidate::PVTight) || lPack->fromPV() == (pat::PackedCandidate::PVLoose)){ 
           pReco.id = 0;
-          if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) < fDZCut)) pReco.id = 1;
-          if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) > fDZCut)) pReco.id = 2;
-          if (fPuppiForLeptons && lPack->fromPV() == (pat::PackedCandidate::PVLoose)) pReco.id = 2;
-          if (fPuppiForLeptons && lPack->fromPV() == (pat::PackedCandidate::PVTight)) pReco.id = 1;
+            if ((fPtMaxCharged > 0) and (pReco.pt > fPtMaxCharged))
```
Comment (perrotta) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: Indentation

File: CommonTools/PileupAlgos/plugins/PuppiProducer.h [hunk 4]
```diff
@@ -48,9 +48,12 @@ class PuppiProducer : public edm::stream::EDProducer<> {
 	std::string     fPVName;
 	bool 			fPuppiDiagnostics;
 	bool 			fPuppiForLeptons;
+  bool fUseFromPVLooseTight;
```
Comment (perrotta) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.h]: Indentation

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 5]
```diff
@@ -16,6 +16,7 @@ PuppiContainer::PuppiContainer(const edm::ParameterSet &iConfig) {
     fUseExp          = iConfig.getParameter<bool>("useExp");
     fPuppiWeightCut  = iConfig.getParameter<double>("MinPuppiWeight");
     fPtMax           = iConfig.getParameter<double>("PtMaxNeutrals");
+  fPtMaxNeutralsStartSlope = iConfig.getParameter<double>("PtMaxNeutralsStartSlope");
```
Comment (perrotta) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc]: Indentation

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 6]
```diff
@@ -269,7 +270,11 @@ std::vector<double> const & PuppiContainer::puppiWeights() {
         }
         //Basic Cuts
         if(pWeight*fPFParticles[i0].pt()   < fPuppiAlgo[pPupId].neutralPt(fNPV) && rParticle.id == 0 ) pWeight = 0;  //threshold cut on the neutral Pt
-        if((fPtMax>0) && (rParticle.id == 0)) pWeight=min(max(pWeight,fPFParticles[i0].pt()/fPtMax),1.);
+        if((fPtMax>0) && (rParticle.id == 0))
+      pWeight =
```
Comment (perrotta) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc]: Indentation

**Review by perrotta — COMMENTED** [review]
The puppi producer code diverged already from 10_6 to the master version
At a glance, the updates here are applied consistently with what done in the master PR #28558.
Did you validate the resukts in both releases, just to confirm that they are consistent?

File: CommonTools/PileupAlgos/interface/PuppiContainer.h (hunk 1 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/interface/PuppiContainer.h]: done

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 2 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: done

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 3 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: done

File: CommonTools/PileupAlgos/plugins/PuppiProducer.h (hunk 4 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/plugins/PuppiProducer.h]: done

File: CommonTools/PileupAlgos/src/PuppiContainer.cc (hunk 5 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/src/PuppiContainer.cc]: done

File: CommonTools/PileupAlgos/src/PuppiContainer.cc (hunk 6 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/src/PuppiContainer.cc]: done

**Comment by ahinzmann** [comment]
I made sure that the diff has the same code as in the master PR. The effect of the PR is the same (no changes to RECO).
The backported code will be validated once more, when we activate the new options for JMECustomNanoAODs for UL and for Run3 afterwards.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Updates functionally identical to the ones implemented in the backported PR #28558 submitted for the master (while the code for the puppiProducer started to diverge underneath)
- A new parameterization is made available for puppi weights, as planned: justification for it is in the attached presentation, while a full large stat validation will follow.
- puppiNoLepton updated to better corresponds to its name
- Modified configs are not activated in the current workflows: no change in reco or miniAOD outputs onserved
- Jenkins tests pass

**Comment by santocch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**